### PR TITLE
[azure-storage-cpp] Added cmake config & targets

### LINKED ISSUE #18781: [azure-storage-cpp] Add CMake config & targets
****

**Is your feature request related to a problem? Please describe.**
It would be nice to have CMake config & targets and move away from referencing the libs, headers and defines directly.

**Proposed solution**
Add CMake config & targets

**Describe alternatives you've considered**
N/A

**Additional context**
The original repo is in maintenance mode with possibly no further releases. Any work to add the CMake config & targets would therefore have to happen on the vcpkg side, even if it's really feature work.

### PR DESCRIPTION
**Pull request by bobuc-havok**
**Describe the pull request**

- #### What does your PR fix?  
  Fixes #18781 

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all previously supported, No

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

**Comment by ahsonkhan** [comment]
cc @Jinming-Hu

**Comment by NancyLi1013** [comment]
The failures on x64-linux and x64-osx like this:

```
CMake Error at /agent/_work/1/s/scripts/buildsystems/vcpkg.cmake:792 (_find_package):
  Could not find a package configuration file provided by "libuuid" with any
  of the following names:

    libuuidConfig.cmake
    libuuid-config.cmake

  Add the installation prefix of "libuuid" to CMAKE_PREFIX_PATH or set
  "libuuid_DIR" to a directory containing one of the above files.  If
  "libuuid" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  CMakeLists.txt:57 (find_package)
```

**Review by NancyLi1013 — COMMENTED** [review]
Could you please help submit these changes to upstream?

**Comment by bucurb** [comment]
@NancyLi1013 build should be fixed now, sorry for the delay. I was waiting for the libuuid targets PR to drop. Regarding helping merge the change upstream, I think that's a lost battle, since the upstream is in maintenance only mode and they're only releasing on critical security fixes. Unfortunately, the new azure SDK does not have the same features as this SDK so we can't just abandon this one in favor of the new one either...

File: ports/azure-storage-cpp/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "azure-storage-cpp",
+  "version-string": "7.5.0",
```
Comment (NancyLi1013) [inline ports/azure-storage-cpp/vcpkg.json:3]:
```suggestion
  "version-semver": "7.5.0",
```
This should be served by [version-semver](https://github.com/microsoft/vcpkg/blob/master/docs/users/versioning.md#version-semver).

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
@Jinming-Hu Could you please help review the PR? Thanks!

**Comment by Jinming-Hu** [comment]
I have a question, I see the following output when I finished installing `azure-storage-cpp`

```
Building package azure-storage-cpp[core]:x64-windows... done
Installing package azure-storage-cpp[core]:x64-windows...
Installing package azure-storage-cpp[core]:x64-windows... done
Elapsed time for package azure-storage-cpp:x64-windows: 1.305 min

Total elapsed time: 2.204 min

The package azure-storage-cpp:x64-windows provides CMake targets:

    find_package(unofficial-azure-storage-cpp CONFIG REQUIRED)
    target_link_libraries(main PRIVATE unofficial::azurestorage)
```


What does the `unofficial` mean here?

**Comment by NancyLi1013** [comment]
> I have a question, I see the following output when I finished installing `azure-storage-cpp`
> 
> ```
> Building package azure-storage-cpp[core]:x64-windows... done
> Installing package azure-storage-cpp[core]:x64-windows...
> Installing package azure-storage-cpp[core]:x64-windows... done
> Elapsed time for package azure-storage-cpp:x64-windows: 1.305 min
> 
> Total elapsed time: 2.204 min
> 
> The package azure-storage-cpp:x64-windows provides CMake targets:
> 
>     find_package(unofficial-azure-storage-cpp CONFIG REQUIRED)
>     target_link_libraries(main PRIVATE unofficial::azurestorage)
> ```
> 
> What does the `unofficial` mean here?

`unofficial` means this is not provided by upstream. We can submit the changes to upstream if this is accepted by upstream. Once merged to upstream, we can remove the prefix `unofficial`.

**Review by Jinming-Hu — APPROVED** [review]
LGTM

File: ports/azure-storage-cpp/vcpkg.json (hunk 1 above)
Comment (bucurb) [reply ports/azure-storage-cpp/vcpkg.json]: done

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
LGTM now, thanks for your PR @bucurb. Also thanks for your help to review @Jinming-Hu.

**Review by BillyONeal — CHANGES_REQUESTED** [review]
I think whether definitions are PUBLIC needs to be more thoughtfully considered

**Comment by NancyLi1013** [comment]
@bucurb Could you please address the review suggestion?

**Comment by bucurb** [comment]
@NancyLi1013 sorry I won't have time to look after this PR for a while.

**Comment by ahsonkhan** [comment]
> Unfortunately, the new azure SDK does not have the same features as this SDK so we can't just abandon this one in favor of the new one either...

@bucurb can you share what are some features that the new SDK is missing, that you need and depend on?

**Comment by Cheney-W** [comment]
Pinging @bucurb  for response. Is work still being done for this PR?

**Comment by Cheney-W** [comment]
Closing this PR since it seems that no progress is being made. Please reopen if work is still being done.

**Pull request closed without merging**